File: packages/omnia-router/src/graph.ts
import type { ChannelGraph, ChannelGraphEdge, SwapAnnouncement } from './types.js';
import { parseRateToScaled } from './pathfind.js';

/** Create an empty ChannelGraph. */
export function createChannelGraph(): ChannelGraph {
  return {
    nodeEdges:       new Map(),
    edgesByChannel:  new Map(),
    swapIndex:       new Map(),
  };
}

/**
 * Add a directed channel edge to the graph.
 *
 * A single logical channel between two parties has two directed edges — one
 * per direction of flow.  Both may be added independently using the same
 * `channelId` because they are keyed by `(channelId, from)`.  Calling
 * `addChannel` twice with the same `(channelId, from)` replaces the first
 * entry (balance update semantics).
 */
export function addChannel(graph: ChannelGraph, edge: ChannelGraphEdge): void {
  const existing = graph.edgesByChannel.get(edge.channelId) ?? [];

  // Replace any edge for the same (channelId, from) direction.
  const idx = existing.findIndex(e => e.from === edge.from);
  if (idx >= 0) {
    _removeFromNodeEdges(graph, existing[idx]);
    existing.splice(idx, 1, edge);
  } else {
    existing.push(edge);
  }
  graph.edgesByChannel.set(edge.channelId, existing);

  // Register in nodeEdges for O(1) Dijkstra lookup.
  const nodeList = graph.nodeEdges.get(edge.from) ?? [];
  nodeList.push(edge);
  graph.nodeEdges.set(edge.from, nodeList);
}

/**
 * Remove ALL directed edges for `channelId` (i.e. both directions).
 * No-op if the channelId is not present.
 */
export function removeChannel(graph: ChannelGraph, channelId: string): void {
  const edges = graph.edgesByChannel.get(channelId);
  if (!edges) return;
  for (const edge of edges) {
    _removeFromNodeEdges(graph, edge);
  }
  graph.edgesByChannel.delete(channelId);
}

function _removeFromNodeEdges(graph: ChannelGraph, edge: ChannelGraphEdge): void {
  const nodeList = graph.nodeEdges.get(edge.from);
  if (!nodeList) return;
  const filtered = nodeList.filter(e => !(e.channelId === edge.channelId && e.from === edge.from));
  if (filtered.length === 0) {
    graph.nodeEdges.delete(edge.from);
  } else {
    graph.nodeEdges.set(edge.from, filtered);
  }
}

/**
 * Register a swap announcement from a bridging intermediary.
 *
 * Validates that the rate is a positive finite decimal before storing.
 * Throws `Error` if the rate is zero or negative.
 *
 * Duplicate announcements (same `intermediaryPubKey` + `inboundChannelId`)
 * are replaced to prevent stale rate data.
 */
export function announceSwap(graph: ChannelGraph, announcement: SwapAnnouncement): void {
  const rateScaled = parseRateToScaled(announcement.rate);
  if (rateScaled <= 0n) {
    throw new Error(
      `Invalid swap rate "${announcement.rate}" from intermediary ${announcement.intermediaryPubKey}: rate must be positive`,
    );
  }

  const key  = _swapKey(announcement.tokenIn, announcement.tokenOut);
  const list = graph.swapIndex.get(key) ?? [];
  const idx  = list.findIndex(
    a => a.intermediaryPubKey === announcement.intermediaryPubKey &&
         a.inboundChannelId   === announcement.inboundChannelId,
  );
  if (idx >= 0) {
    list[idx] = announcement;
  } else {
    list.push(announcement);
  }
  graph.swapIndex.set(key, list);
}

/**
 * Return all swap announcements for a given token pair, or an empty array.
 */
export function getSwapAnnouncements(
  graph:    ChannelGraph,
  tokenIn:  string,
  tokenOut: string,
): SwapAnnouncement[] {
  return graph.swapIndex.get(_swapKey(tokenIn, tokenOut)) ?? [];
}

function _swapKey(tokenIn: string, tokenOut: string): string {
  return `${tokenIn}:${tokenOut}`;
}

File: packages/omnia-router/go/graph.go
package omniarouter

import (
	"math/big"
)

func (g *ChannelGraph) AddChannel(edge ChannelGraphEdge) {
	existing := g.EdgesByChannel[edge.ChannelID]

	idx := -1
	for i, e := range existing {
		if e.From == edge.From {
			idx = i
			break
		}
	}
	if idx >= 0 {
		g.removeFromNodeEdges(existing[idx])
		existing[idx] = edge
	} else {
		existing = append(existing, edge)
	}
	g.EdgesByChannel[edge.ChannelID] = existing

	nodeList := g.NodeEdges[edge.From]
	nodeList = append(nodeList, edge)
	g.NodeEdges[edge.From] = nodeList
}

func (g *ChannelGraph) RemoveChannel(channelID string) {
	edges, ok := g.EdgesByChannel[channelID]
	if !ok {
		return
	}
	for _, edge := range edges {
		g.removeFromNodeEdges(edge)
	}
	delete(g.EdgesByChannel, channelID)
}

func (g *ChannelGraph) removeFromNodeEdges(edge ChannelGraphEdge) {
	nodeList := g.NodeEdges[edge.From]
	filtered := make([]ChannelGraphEdge, 0, len(nodeList))
	for _, e := range nodeList {
		if !(e.ChannelID == edge.ChannelID && e.From == edge.From) {
			filtered = append(filtered, e)
		}
	}
	if len(filtered) == 0 {
		delete(g.NodeEdges, edge.From)
	} else {
		g.NodeEdges[edge.From] = filtered
	}
}

func (g *ChannelGraph) AnnounceSwap(announcement SwapAnnouncement) error {
	rateScaled := ParseRateToScaled(announcement.Rate)
	if rateScaled.Sign() <= 0 {
		return &SwapError{
			Intermediary: announcement.IntermediaryPubKey,
			Rate:         announcement.Rate,
		}
	}

	key := swapKey(announcement.TokenIn, announcement.TokenOut)
	list := g.SwapIndex[key]

	idx := -1
	for i, a := range list {
		if a.IntermediaryPubKey == announcement.IntermediaryPubKey &&
			a.InboundChannelID == announcement.InboundChannelID {
			idx = i
			break
		}
	}
	if idx >= 0 {
		list[idx] = announcement
	} else {
		list = append(list, announcement)
	}
	g.SwapIndex[key] = list
	return nil
}

func (g *ChannelGraph) GetSwapAnnouncements(tokenIn, tokenOut string) []SwapAnnouncement {
	return g.SwapIndex[swapKey(tokenIn, tokenOut)]
}

func swapKey(tokenIn, tokenOut string) string {
	return tokenIn + ":" + tokenOut
}

type SwapError struct {
	Intermediary string
	Rate         string
}

func (e *SwapError) Error() string {
	return "Invalid swap rate \"" + e.Rate + "\" from intermediary " + e.Intermediary + ": rate must be positive"
}

func edgeFee(amount *big.Int, edge ChannelGraphEdge) *big.Int {
	feeRate, _ := new(big.Int).SetString(edge.FeeRate, 10)
	result := new(big.Int).Mul(amount, feeRate)
	result.Div(result, big.NewInt(Scale))
	return result
}
